gadget/gadget.go

```diff
@@ -299,41 +329,106 @@ func validateVolume(name string, vol *Volume) error {
 	}
 
-	// validate relative offsets that reference other structures by name
-	for idx, s := range vol.Structure {
-		if s.OffsetWrite.RelativeTo != "" && !structureNames[s.OffsetWrite.RelativeTo] {
-			return fmt.Errorf("structure %v refers to an unknown structure %q",
-				fmtIndexAndName(idx, s.Name), s.OffsetWrite.RelativeTo)
+	// sort by starting offset
+	sort.Sort(byStartOffset(structures))
+
+	return validateCrossVolumeStructure(structures, knownStructures)
+}
+
+func validateCrossVolumeStructure(structures []PositionedStructure, knownStructures map[string]*PositionedStructure) error {
+	previousEnd := Size(0)
+	// cross structure validation:
+	// - relative offsets that reference other structures by name
+	// - positioned structure overlap
+	// use structures positioned within the volume
+	for pidx, ps := range structures {
+		if ps.Role == "mbr" || ps.Type == "mbr" {
+			if ps.StartOffset != 0 {
+				return fmt.Errorf(`structure %v has "mbr" role and must start at offset 0`, ps)
+			}
+		}
+		if ps.OffsetWrite != nil && ps.OffsetWrite.RelativeTo != "" {
+			// offset-write using a named structure
+			other := knownStructures[ps.OffsetWrite.RelativeTo]
+			if other == nil {
+				return fmt.Errorf("structure %v refers to an unknown structure %q",
+					ps, ps.OffsetWrite.RelativeTo)
+			}
+			// offset is written as a 4-byte pointer value at offset-write address
+			if ps.OffsetWrite.Offset > (other.Size - SizeLBA48Pointer) {
+				return fmt.Errorf("structure %v offset-write crosses structure %v size",
+					ps, other)
+			}
+
 		}
 
-		if !s.IsBare() {
+		if ps.StartOffset < previousEnd {
+			previous := structures[pidx-1]
+			return fmt.Errorf("structure %v overlaps with the preceding structure %v", ps, previous)
+		}
+		previousEnd = ps.StartOffset + ps.Size
+
+		if !ps.IsBare() {
 			// content relative offset only possible if it's a bare structure
 			continue
 		}
-		for cidx, c := range s.Content {
-			if c.OffsetWrite.RelativeTo != "" && !structureNames[c.OffsetWrite.RelativeTo] {
+		for cidx, c := range ps.Content {
+			if c.OffsetWrite == nil || c.OffsetWrite.RelativeTo == "" {
+				continue
+			}
+			relativeToStructure := knownStructures[c.OffsetWrite.RelativeTo]
+			if relativeToStructure == nil {
 				return fmt.Errorf("structure %v, content %v refers to an unknown structure %q",
-					fmtIndexAndName(idx, s.Name), fmtIndexAndName(cidx, c.Image), c.OffsetWrite.RelativeTo)
+					ps, fmtIndexAndName(cidx, c.Image), c.OffsetWrite.RelativeTo)
+			}
+			// offset is written as a 4-byte pointer value at offset-write address
+			if c.OffsetWrite.Offset > (relativeToStructure.Size - SizeLBA48Pointer) {
+				return fmt.Errorf("structure %v, content %v offset-write crosses structure %q size",
+					ps, fmtIndexAndName(cidx, c.Image), c.OffsetWrite.RelativeTo)
 			}
 		}
 	}
 	return nil
 }
 
 func validateVolumeStructure(vs *VolumeStructure, vol *Volume) error {
+	if vs.Size == 0 {
+		return errors.New("missing size")
```

Comment:
Does this look OK in the final composed error?

Reply:
It's pretty specific: `invalid volume "foo": invalid structure #<idx> (<name>): missing size`